<a href="https://colab.research.google.com/github/LFagundesN/PythonClass/blob/main/A1_Student_Scores_Python_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **A1: Student Scores Python Assignment**

### Leticia Fagundes Netto

GitHub: https://github.com/LFagundesN/PythonClass/blob/270c8d7b2eeef6429b02cd5b3207b55bbcaf1417/A1_Student_Scores_Python_Assignment.ipynb

Write a short Python script that meets all the following requirements:

Your code must:

*  Define at least one function that performs a specific task.
*  Create and use a dictionary to store or process information.
*  Include a loop (either for or while) to iterate over data.
*  Use a conditional (if, elif, else) to make decisions.

In [19]:
# STUDENT DATA GENERATION SCRIPT – EXPLAINED
# Each student is assigned:
#   A unique student ID (e.g. 1, 2, ..., 50)
#   A name generated by ChatGPT
#   Two grades (Grade 1 and Grade 2) generated randomly within realistic ranges:
#     • Grade 1 → random integer between 40 and 100
#     • Grade 2 → random integer between 30 and 100

# >>> CODE USED TO GENERATE IT:
#
import random

# List of 50 random student names (generated via ChatGPT)

names = [
    "Olivia Carter", "Ethan Brooks", "Sophia Bennett", "Noah Wallace", "Isabella Grant", "Liam Turner",
    "Amelia Foster", "Mason Clarke", "Ava Reynolds", "Lucas Hayes", "Mia Saunders", "Elijah Ford",
    "Charlotte Dean", "Henry Walsh", "Emily Lawson", "Benjamin Cole", "Grace Mitchell", "James Porter",
    "Lily Hart", "Daniel Webb", "Chloe Adams", "Alexander Drake", "Harper Quinn", "Jacob Palmer",
    "Ella Maxwell", "Samuel Bates", "Scarlett James", "Leo Armstrong", "Ivy Newman", "Caleb Ross",
    "Aria Blake", "Nathan Ellis", "Zoe Chapman", "Gabriel Lane", "Ruby Sinclair", "Isaac Doyle",
    "Nora Matthews", "Owen Barrett", "Isla Hudson", "Julian Shepherd", "Maya Gilbert", "Dylan Rhodes",
    "Sienna Fraser", "Aiden Lowe", "Violet Morgan", "Ryan Bell", "Penelope Hughes", "Austin McCarthy",
    "Eliza Rowe", "Theo Nash"
]

# Function to generate student records with random grades
def generate_student_data():
    student_data = []
    for i, name in enumerate(names):
        student_id = f"{i+1}"                   # Generate sequential ID (e.g., St_1)
        grade1 = random.randint(40, 100)            # Grade 1: realistic score range
        grade2 = random.randint(30, 100)            # Grade 2: realistic score range
        student_data.append((student_id, name, grade1, grade2))  # Store as tuple
    return student_data

# Generate the dataset
students = generate_student_data()

# Display all 50 entries (Student ID, Name, Grade 1, Grade 2)
print("Generated Student Data (St_id, Name, Grade 1, Grade 2):\n")
for student in students:
    print(student)

Generated Student Data (St_id, Name, Grade 1, Grade 2):

('1', 'Olivia Carter', 81, 32)
('2', 'Ethan Brooks', 40, 82)
('3', 'Sophia Bennett', 53, 73)
('4', 'Noah Wallace', 57, 45)
('5', 'Isabella Grant', 72, 38)
('6', 'Liam Turner', 52, 48)
('7', 'Amelia Foster', 40, 65)
('8', 'Mason Clarke', 63, 52)
('9', 'Ava Reynolds', 69, 83)
('10', 'Lucas Hayes', 84, 37)
('11', 'Mia Saunders', 69, 32)
('12', 'Elijah Ford', 52, 86)
('13', 'Charlotte Dean', 62, 47)
('14', 'Henry Walsh', 51, 56)
('15', 'Emily Lawson', 74, 72)
('16', 'Benjamin Cole', 60, 92)
('17', 'Grace Mitchell', 95, 86)
('18', 'James Porter', 79, 62)
('19', 'Lily Hart', 99, 66)
('20', 'Daniel Webb', 49, 63)
('21', 'Chloe Adams', 62, 67)
('22', 'Alexander Drake', 57, 92)
('23', 'Harper Quinn', 61, 34)
('24', 'Jacob Palmer', 76, 87)
('25', 'Ella Maxwell', 66, 89)
('26', 'Samuel Bates', 71, 95)
('27', 'Scarlett James', 45, 40)
('28', 'Leo Armstrong', 57, 61)
('29', 'Ivy Newman', 95, 43)
('30', 'Caleb Ross', 41, 67)
('31', 'Aria Blake

In [20]:
# Student Dataset (Dictionary Format)
# This dictionary contains a fixed set of student data.
# Each student is represented by a unique ID (e.g., '1').
# The value for each ID is another dictionary holding:
# - 'name': Full student name
# - 'grade1': Grade from the first assessment
# - 'grade2': Grade from the second assessment

# The original list was generated and then edited manually to:
# - Fix outlier or unrealistic data points
# - Create a more balanced sample (e.g., diversity in grades)
# - Ensure names and scores are aligned with assignment needs

students = {
    '1': {'name': 'Olivia Carter', 'grade1': 77, 'grade2': 89},
    '2': {'name': 'Ethan Brooks', 'grade1': 61, 'grade2': 46},
    '3': {'name': 'Sophia Bennett', 'grade1': 89, 'grade2': 59},
    '4': {'name': 'Noah Wallace', 'grade1': 87, 'grade2': 81},
    '5': {'name': 'Isabella Grant', 'grade1': 73, 'grade2': 85},
    '6': {'name': 'Liam Turner', 'grade1': 76, 'grade2': 94},
    '7': {'name': 'Amelia Foster', 'grade1': 74, 'grade2': 50},
    '8': {'name': 'Mason Clarke', 'grade1': 64, 'grade2': 79},
    '9': {'name': 'Ava Reynolds', 'grade1': 90, 'grade2': 87},
    '10': {'name': 'Lucas Hayes', 'grade1': 76, 'grade2': 40},
    '11': {'name': 'Mia Saunders', 'grade1': 52, 'grade2': 88},
    '12': {'name': 'Elijah Ford', 'grade1': 79, 'grade2': 70},
    '13': {'name': 'Charlotte Dean', 'grade1': 92, 'grade2': 98},
    '14': {'name': 'Henry Walsh', 'grade1': 81, 'grade2': 87},
    '15': {'name': 'Emily Lawson', 'grade1': 80, 'grade2': 77},
    '16': {'name': 'Benjamin Cole', 'grade1': 83, 'grade2': 77},
    '17': {'name': 'Grace Mitchell', 'grade1': 69, 'grade2': 57},
    '18': {'name': 'James Porter', 'grade1': 78, 'grade2': 99},
    '19': {'name': 'Lily Hart', 'grade1': 87, 'grade2': 75},
    '20': {'name': 'Daniel Webb', 'grade1': 67, 'grade2': 82},
    '21': {'name': 'Chloe Adams', 'grade1': 59, 'grade2': 68},
    '22': {'name': 'Alexander Drake', 'grade1': 46, 'grade2': 76},
    '23': {'name': 'Harper Quinn', 'grade1': 60, 'grade2': 30},
    '24': {'name': 'Jacob Palmer', 'grade1': 97, 'grade2': 85},
    '25': {'name': 'Ella Maxwell', 'grade1': 56, 'grade2': 41},
    '26': {'name': 'Samuel Bates', 'grade1': 64, 'grade2': 49},
    '27': {'name': 'Scarlett James', 'grade1': 84, 'grade2': 87},
    '28': {'name': 'Leo Armstrong', 'grade1': 81, 'grade2': 57},
    '29': {'name': 'Ivy Newman', 'grade1': 100, 'grade2': 79},
    '30': {'name': 'Caleb Ross', 'grade1': 75, 'grade2': 54},
    '31': {'name': 'Aria Blake', 'grade1': 59, 'grade2': 32},
    '32': {'name': 'Nathan Ellis', 'grade1': 100, 'grade2': 79},
    '33': {'name': 'Zoe Chapman', 'grade1': 51, 'grade2': 72},
    '34': {'name': 'Gabriel Lane', 'grade1': 75, 'grade2': 48},
    '35': {'name': 'Ruby Sinclair', 'grade1': 66, 'grade2': 84},
    '36': {'name': 'Isaac Doyle', 'grade1': 88, 'grade2': 100},
    '37': {'name': 'Nora Matthews', 'grade1': 65, 'grade2': 86},
    '38': {'name': 'Owen Barrett', 'grade1': 92, 'grade2': 44},
    '39': {'name': 'Isla Hudson', 'grade1': 64, 'grade2': 55},
    '40': {'name': 'Julian Shepherd', 'grade1': 91, 'grade2': 79},
    '41': {'name': 'Maya Gilbert', 'grade1': 87, 'grade2': 42},
    '42': {'name': 'Dylan Rhodes', 'grade1': 76, 'grade2': 70},
    '43': {'name': 'Sienna Fraser', 'grade1': 64, 'grade2': 85},
    '44': {'name': 'Aiden Lowe', 'grade1': 63, 'grade2': 90},
    '45': {'name': 'Violet Morgan', 'grade1': 60, 'grade2': 36},
    '46': {'name': 'Ryan Bell', 'grade1': 68, 'grade2': 86},
    '47': {'name': 'Penelope Hughes', 'grade1': 80, 'grade2': 92},
    '48': {'name': 'Austin McCarthy', 'grade1': 59, 'grade2': 74},
    '49': {'name': 'Eliza Rowe', 'grade1': 78, 'grade2': 93},
    '50': {'name': 'Theo Nash', 'grade1': 83, 'grade2': 74}
}

# Code and formatting assistance was provided by ChatGPT.

In [21]:
# Calculate Final Grades Using Custom Weighting
# For each student, calculate a final grade using the formula:
# final_grade = (grade1 + 2 * grade2) / 3

final_grades = {}

for student_id, data in students.items(): # data is the dictionary of all students.
    g1 = data['grade1']
    g2 = data['grade2']
    final = round((g1 + 2 * g2) / 3, 2)
    # Insert a new key 'final_grade' into that student's data
    students[student_id]['final_grade'] = final

# data are the values of each item in the students dictionary.
# data example:  {'name': 'Olivia Carter', 'grade1': 77, 'grade2': 89}

In [23]:
# Access and print the data for student with ID '1'
print(students['1'])

{'name': 'Olivia Carter', 'grade1': 77, 'grade2': 89, 'final_grade': 85.0}


## Average score

In [31]:
# Function to calculate the average of grade1, grade2, and final_grade
def calculate_averages(students):
    # Total sum of all grades
    total_grade1 = 0
    total_grade2 = 0
    total_final = 0

    count = len(students)  # Number of students in the dataset

# Loop through each student and add to each of their own totals
    for student_id, data in students.items():
        total_grade1 += data['grade1']
        total_grade2 += data['grade2']
        total_final += data['final_grade']

# Calculate averages by dividing totals by number of students
    avg_grade1 = round(total_grade1 / count, 2)
    avg_grade2 = round(total_grade2 / count, 2)
    avg_final = round(total_final / count, 2)

# Return all three averages in a dictionary
    return {
        'average_grade1': avg_grade1,
        'average_grade2': avg_grade2,
        'average_final_grade': avg_final
    }

# Return to send data back from the function.

In [32]:
# Call the function and print the result
averages = calculate_averages(students)
print(averages)

{'average_grade1': 74.52, 'average_grade2': 71.34, 'average_final_grade': 72.4}


The students performed slightly better in Grade 1 (average: 74.52) than in Grade 2 (average: 71.34). This suggests students did better on the first assignment. The overall final grade average sits at 72.4, indicating a generally strong performance across both assessments.

## Highest score

In [67]:
# Function to Find the top 5 students with the highest grade1

def highest_grade1 (data):
  # Sort students by grade1 (highest first), get top 5
  top_g1 = sorted(data.items(), key=lambda x: x[1]['grade1'], reverse=True)[:5]

  # Return the top 5 students for grade1 their name and grade
  return [f"{s[1]['name']} - {s[1]['grade1']}" for s in top_g1]
  # Format each result as a string: "Name - Grade"

# Breakdown of: sorted(data.items(), key=lambda x: x[1]['grade1'], reverse=True)[:5]

# data.items() -> returns all key-value pairs in the dictionary:
# Each item is a tuple: (student_id, student_data)

# key=lambda x: x[1]['grade1'] -> tells sorted() to sort by the student's 'grade1' value.

# Explanation based on Emonds (2023) and Pieters, Martijn & Flakes (2022) on Stack Overflow:
# lambda is used to define an anonymous (unnamed) function.
# Instead of doing this:
#     def grade1_list(x):
#         return x[1]['grade1']
# And then passing: sorted(data.items(), key=grade1_list)
# I used lambda directly:
#     sorted(data.items(), key=lambda x: x[1]['grade1'])

# reverse=True -> sort in descending order (highest grades come first)
# [:5] → keep only the first 5 results (top 5 students by grade1)

In [70]:
# Call the function and print the results
print(highest_grade1(students))

['Ivy Newman - 100', 'Nathan Ellis - 100', 'Jacob Palmer - 97', 'Charlotte Dean - 92', 'Owen Barrett - 92']


In [65]:
print("Top 5 Students with Highest Grade 1:")
for line1 in highest_grade1(students):
    print(line1)

# This code go through each student in the top 5 list from the function
# Print each student one by one (just their ID and grade1)
# Why I did this?
# To print each tuple (student ID and their grade1 score) on a new line.
# All about the format of the output.

Top 5 Students with Highest Grade 1:
Ivy Newman - 100
Nathan Ellis - 100
Jacob Palmer - 97
Charlotte Dean - 92
Owen Barrett - 92


In [68]:
# Function to Find the top 5 students with the highest grade2
# Same as the previous function (highest_grade1)

def highest_grade2 (data):
  # Sort all students by grade2 in descending order and take top 5
  top_g2 = sorted(data.items(), key=lambda x: x[1]['grade2'], reverse=True)[:5]

  # Return the top 5 students for grade1 their name and grade
  return [f"{s[1]['name']} - {s[1]['grade2']}" for s in top_g2]
  # Format each result as a string: "Name - Grade"

In [69]:
# Call the function and print the results
print("Top 5 Students with Highest Grade 2:")
for line2 in highest_grade2(students):
    print(line2)

# Again, I could also use this:
# print(highest_grade1(students))
# But I wanted a better output format.

Top 5 Students with Highest Grade 2:
Isaac Doyle - 100
James Porter - 99
Charlotte Dean - 98
Liam Turner - 94
Eliza Rowe - 93


In [71]:
# Function to Find the top 5 students with the highest final_grade

def highest_final (data):
  # Sort all students by final_grade in descending order and take top 5
  top_final = sorted(data.items(), key=lambda x: x[1]['final_grade'], reverse=True)[:5]

  # Return the top 5 students for grade1 their name and grade
  return [f"{s[1]['name']} - {s[1]['final_grade']}" for s in top_final]
  # Format each result as a string: "Name - Grade"

In [72]:
# Call the function and print the results
print("Top 5 Students with Highest Final Grades:")
for linef in highest_final(students):
    print(linef)

# Again, I could also use this:
# print(highest_grade1(students))
# But I wanted a better output format.

Top 5 Students with Highest Final Grades:
Charlotte Dean - 96.0
Isaac Doyle - 96.0
James Porter - 92.0
Jacob Palmer - 89.0
Liam Turner - 88.0


Charlotte Dean appears across all three top-5 lists, showing balanced strength in both individual assessments and final performance.
<br>
Isaac Doyle, James Porter and Liam Turner also rank in grade 2 and final scores, highlighting strong finishes even if not top in grade 1.
<br>
Ivy Newman and Nathan Ellis lead in grade 1 with perfect scores but don’t appear in final grade rankings, possibly due to lower second grades.

## Number of students who passed (score ≥ 50)

Grading System Classification by Final Grade:

*   Passed with Distinction: >85
*   Passed with Merit: 70–85 (inclusive)
*   Passed: 50-69
*   Failed: <50


In [77]:
# Function: Classify students by final grade category
def classify_students(data):

    # Start counters for each grade category
    distinction = 0  # >85
    merit = 0        # 70–85
    passed = 0       # 50–69
    failed = 0       # <50

    # Loop through all students
    for student_id, info in data.items():
        final = info['final_grade']  # Get their final grade

        # Check which grade category they fall into
        if final > 85:
            distinction += 1
        elif 70 <= final <= 85:
            merit += 1
        elif 50 <= final < 70:
            passed += 1
        else:
            failed += 1

    # Return the total count for each category
    return distinction, merit, passed, failed

In [79]:
# Run the classification function and store the results
# d = distinction, m = merit, p = passed, f = failed

d, m, p, f = classify_students(students)

# Print the number of students in each classification
print("Distinction:", d)
print("Merit:", m)
print("Passed:", p)
print("Failed:", f)

Distinction: 11
Merit: 19
Passed: 16
Failed: 4


In [82]:
# Calculate total number of students who passed (Distinction + Merit + Passed)
TotalPassed = sum([d, m, p])

# Total number of students who failed (final grade < 50)
TotalFailed = f

# Total number of students overall (Passed + Failed) = 50
TotalStudents = TotalPassed + TotalFailed

# Print total number of students in each category
print(f"Total Passed: {TotalPassed}")
print(f"Total Failed: {TotalFailed}")

# Calculate percentage of students who passed
Pct_Passed = round((TotalPassed / TotalStudents) * 100, 2)

# Calculate percentage of students who failed
Pct_Failed = round((TotalFailed / TotalStudents) * 100, 2)

# Print percentage results
print(f"Percentage Passed: {Pct_Passed}%")
print(f"Percentage Failed: {Pct_Failed}%")

# The f"" (f-string):
# Easier way to output variables.
# Cleaner and better output format

Total Passed: 46
Total Failed: 4
Percentage Passed: 92.0%
Percentage Failed: 8.0%


Out of 50 students, 46 passed and only 4 failed. Most students did well. 11 got Distinction (above 85), 19 got Merit (70–85), and 16 just Passed (50–69). Only 4 scored below 50. So, 92% passed and 8% failed. Overall, strong performance.

## How GPT was used

ChatGPT was used to help structure the dataset, correct and comment functions, explain code logic, and summarise the results. It played a key role in writing clear, detailed comments and in breaking down the overall logic of the project in a simple way.

## References

OpenAI. (2024). *ChatGPT* (May 2024 version) [Large language model]. https://chat.openai.com/

Emonds, C. (2023, March 7). Answer to: What is lambda in Python code? How does it work with key arguments to sorted, sum etc.? Stack Overflow. https://stackoverflow.com/a/13669252

Pieters, M., & Flakes. (2023). Answer to: What is lambda in Python code? How does it work with key arguments to sorted, sum etc.? Stack Overflow. https://stackoverflow.com/a/13669252

hygull. (2017, September 2). Answer to: What is lambda in Python code? How does it work with key arguments to sorted, sum etc.? Stack Overflow. https://stackoverflow.com/a/46014149

Knechtel, K. (2022, September 5). Edit to: Answer by hygull on “What is lambda in Python code?” Stack Overflow. https://stackoverflow.com/posts/46014149/revisions
